### PR TITLE
add annotations to inject assumptions

### PR DESCRIPTION
**Pull request by o-**
(this is mostly convenience to generate examples. review is not urgent, I wanted to publish it so you can all use it.)

guard_hint is a way of injecting an assumption.

in the assumption pass sourir will promote guard_hint into the nearest
condition for an osr instruction.

The instruction is called guard and not osr_hint because I would like to
rename osr to guard too in the future. More on that later...

### CODE REVIEW

File: analysis.ml [hunk 1]
```diff
@@ -26,7 +26,7 @@ let successors_at (instrs : instructions) resolve pc : pc list =
   let instr = instrs.(pc) in
   let all_succ =
     match instr with
-    | Decl_var _ | Decl_array _
+    | GuardHint _ | Decl_var _ | Decl_array _
```
Comment (mhyee) [inline analysis.ml]: `Guard_hint`

File: test_examples.sh [hunk 2]
```diff
@@ -12,7 +12,7 @@ function ncores {
   fi
 }
 
-export OPTS="prune\nprune_no_hoist\nhoist_osr\nconst_fold\nhoist_assign\nhoist_drop\nmin_live\ninline_small\ninline_med"
+export OPTS="prune\nprune_no_hoist\nassumptions\nhoist_osr\nconst_fold\nhoist_assign\nhoist_drop\nmin_live\ninline_small\ninline_med"
```
Comment (mhyee) [inline test_examples.sh:15]: If we're adding this to the test runner, maybe we should have add a short program to the `examples` directory? The one in `tests.ml` should be enough.

File: tests.ml [hunk 3]
```diff
@@ -1741,6 +1741,55 @@ let do_test_autofix () =
...
+    var x = 1
+    guard_hint (x==1)
+    branch (x==1) $la $lb
+    $la:
+      return 1
+    $lb:
+      return 2
+  |input} {expect|
+    function main ()
+    version anon_1
```
Comment (mhyee) [inline tests.ml:1772]: It looks like two versions are created. One which adds a checkpoint/guard/osr, and then the active version which removes the hint. In fact, it looks like a new version is created for each hint?

I think it'd be better if all the guards were added and hints removed in the same version. That way, there are only two versions: the new/active one, and the old/original one.

**Review by mhyee — COMMENTED** [review]
Looked over this quickly. Merging/addressing comments is low priority.

File: tests.ml (hunk 3 above)
Comment (o-) [reply tests.ml:1772]: that's not sound. in the baseline the checkpoints need to be empty

File: analysis.ml (hunk 1 above)
Comment (o-) [reply analysis.ml]: ups, old habits

File: tests.ml (hunk 3 above)
Comment (mhyee) [reply tests.ml:1772]: I'm not sure I follow. In my proposal, the baseline is the original version, which has no checkpoints.

Comment (o-) [reply tests.ml:1772]: the design decision was to insert the checkpoints in the baseline to be able to create new versions by just copying.

Comment (o-) [reply tests.ml:1772]: all our passes do this

Comment (mhyee) [reply tests.ml:1772]: OK, I see. Never mind.

File: test_examples.sh (hunk 2 above)
Comment (o-) [reply test_examples.sh:15]: I'll do that in a later PR

**Pull request merged**